File: Source/Components/Net/Nequeo.Sip/Nequeo.PjSip/Nequeo.PjSip/VideoPreview.cpp
/* Company :       Nequeo Pty Ltd, http://www.nequeo.com.au/
*  Copyright :     Copyright © Nequeo Pty Ltd 2015 http://www.nequeo.com.au/
*
*  File :          VideoPreview.cpp
*  Purpose :       SIP VideoPreview class.
*
*/

/*
Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.
*/

#include "stdafx.h"

#include "VideoPreview.h"

using namespace Nequeo::Net::PjSip;

/// <summary>
/// Video preview.
/// </summary>
/// <param name="captureDeviceID">The video capture device id.</param>
VideoPreview::VideoPreview(int captureDeviceID) :
	_pjVideoPreview(new pj::VideoPreview(captureDeviceID)), _disposed(false)
{
}

///	<summary>
///	Video preview deconstructor.
///	</summary>
VideoPreview::~VideoPreview()
{
	if (!_disposed)
	{
		// Cleanup the native classes.
		this->!VideoPreview();

		_disposed = true;
	}
}

///	<summary>
///	Video preview.
///	</summary>
VideoPreview::!VideoPreview()
{
	if (!_disposed)
	{
		// If the callback has been created.
		if (_pjVideoPreview != nullptr)
		{
			// Cleanup the native classes.
			delete _pjVideoPreview;
			_pjVideoPreview = nullptr;
		}
	}
}

/// <summary>
/// Determine if the specified video input device has built-in native
/// preview capability.This is a convenience function that is equal to
/// querying device's capability for PJMEDIA_VID_DEV_CAP_INPUT_PREVIEW
/// capability.
/// </summary>
/// <returns>True if it has.</returns>
bool VideoPreview::HasNative()
{
	return _pjVideoPreview->hasNative();
}

/// <summary>
/// Start video preview window for the specified capture device.
/// </summary>
/// <param name="param">Video preview parameters.</param>
void VideoPreview::Start(VideoPreviewOpParam^ param)
{
	pj::VideoPreviewOpParam prm;

	// If not null.
	if (param != nullptr)
	{
		prm.rendId = param->RenderID;
		prm.show = param->Show;
		prm.windowFlags = param->WindowFlags;

		if (param->Format != nullptr)
		{
			prm.format.type = MediaFormat::GetMediaType(param->Format->Type);
			prm.format.id = param->Format->Id;
		}

		if (param->Window != nullptr)
		{
			prm.window.type = VideoWindow::GeVideoDeviceHandleType(param->Window->Type);
		}
	}
	else
	{
		prm.rendId = pjmedia_vid_dev_std_index::PJMEDIA_VID_DEFAULT_RENDER_DEV;
		prm.show = true;
		prm.windowFlags = pjmedia_vid_dev_wnd_flag::PJMEDIA_VID_DEV_WND_BORDER | pjmedia_vid_dev_wnd_flag::PJMEDIA_VID_DEV_WND_RESIZABLE;
	}

	try
	{
		// Start the preview.
		_pjVideoPreview->start(prm);
	}
	catch (const pj::Error& pjError)
	{
		std::string reason = pjError.reason;
		String^ err = gcnew String(reason.c_str());
		throw gcnew System::Exception(err);
	}
	catch (const std::exception& ex) 
	{
		std::string reason = ex.what();
		String^ err = gcnew String(reason.c_str());
		throw gcnew System::Exception(err);
	}
}

/// <summary>
/// Stop video preview.
/// </summary>
void VideoPreview::Stop()
{
	_pjVideoPreview->stop();
}

/// <summary>
/// Get the preview window handle associated with the capture device, if any.
/// </summary>
/// <returns>Video window.</returns>
VideoWindow^ VideoPreview::GetVideoWindow()
{
	pj::VideoWindow pjWindow = _pjVideoPreview->getVideoWindow();
	VideoWindow^ window = gcnew VideoWindow(pjWindow);
	window->SetVideoPreviewRef(_pjVideoPreview);

	// Return the video window.
	return window;
}

File: Source/Clients/Tools/Nequeo.VoIP.PjSip/Nequeo.VoIP.PjSip.Controls/UI/VideoPreview.cs
/*  Company :       Nequeo Pty Ltd, http://www.nequeo.com.au/
 *  Copyright :     Copyright © Nequeo Pty Ltd 2016 http://www.nequeo.com.au/
 * 
 *  File :          
 *  Purpose :       
 * 
 */

#region Nequeo Pty Ltd License
/*
    Permission is hereby granted, free of charge, to any person
    obtaining a copy of this software and associated documentation
    files (the "Software"), to deal in the Software without
    restriction, including without limitation the rights to use,
    copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the
    Software is furnished to do so, subject to the following
    conditions:

    The above copyright notice and this permission notice shall be
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
    OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
    HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
    OTHER DEALINGS IN THE SOFTWARE.
*/
#endregion

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Nequeo.VoIP.PjSip.UI
{
    /// <summary>
    /// Video preview.
    /// </summary>
    public partial class VideoPreview : Form
    {
        /// <summary>
        /// Instant message.
        /// </summary>
        /// <param name="voipCall">VoIP call.</param>
        public VideoPreview(Nequeo.VoIP.PjSip.VoIPCall voipCall)
        {
            InitializeComponent();

            _voipCall = voipCall;
            Create();
        }

        private bool _disposed = false;
        private Nequeo.VoIP.PjSip.VoIPCall _voipCall = null;
        private Nequeo.Net.PjSip.VideoPreview _videoPreview = null;
        private Nequeo.Net.PjSip.VideoWindow _window = null;

        /// <summary>
        /// Form is closing.
        /// </summary>
        public event System.EventHandler OnVideoPreviewClosing;

        /// <summary>
        /// Dispose of the unmanaged resources.
        /// </summary>
        public void DisposeCall()
        {
            // Check to see if Dispose has already been called.
            if (!this._disposed)
            {
                // Note disposing has been done.
                _disposed = true;

                try
                {
                    // If disposing equals true, dispose all managed
                    // and unmanaged resources.
                    if (_window != null)
                        _window.Dispose();
                }
                catch { }

                try
                {
                    // The video preview.
                    if (_videoPreview != null)
                        _videoPreview.Dispose();
                }
                catch { }

                // Call the appropriate methods to clean up
                // unmanaged resources here.
                _window = null;
                _videoPreview = null;
            }
        }

        /// <summary>
        /// Create the preview.
        /// </summary>
        private void Create()
        {
            try
            {
                // Get the current video capture device id.
                int videoCapID = _voipCall.VoIPManager.MediaManager.GetVideoCaptureDeviceID();
                int videoRenID = _voipCall.VoIPManager.MediaManager.GetVideoRenderDeviceID();

                // If a capture device exist.
                if (videoCapID >= -1)
                {
                    // Create.
                    _videoPreview = new Nequeo.Net.PjSip.VideoPreview(videoCapID);

                    // Configure.
                    Nequeo.Net.PjSip.VideoPreviewOpParam parm = new Nequeo.Net.PjSip.VideoPreviewOpParam();

                    Nequeo.Net.PjSip.VideoWindowHandle handle = new Nequeo.Net.PjSip.VideoWindowHandle();
                    handle.Type = Nequeo.Net.PjSip.VideoDeviceHandleType.PJMEDIA_VID_DEV_HWND_TYPE_WINDOWS;

                    Nequeo.Net.PjSip.MediaFormat format = new Nequeo.Net.PjSip.MediaFormat();
                    format.Type = Nequeo.Net.PjSip.MediaType.PJMEDIA_TYPE_VIDEO;

                    // Assign.
                    parm.Show = false;
                    parm.Format = format;
                    parm.WindowFlags = 0;
                    parm.RenderID = videoRenID;
                    parm.Window = handle;

                    // Show.
                    _videoPreview.Start(parm);
                    _window = _videoPreview.GetVideoWindow();
                }
            }
            catch { }
        }

        /// <summary>
        /// Form move.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void VideoPreview_Move(object sender, EventArgs e)
        {
            int top = this.Top + 43;
            int left = this.Left + 20;

            try
            {
                _window.SetPosition(
                    new Nequeo.Net.PjSip.MediaCoordinate()
                    {
                        X = left,
                        Y = top
                    });
            }
            catch { }
        }

        /// <summary>
        /// Form load.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void VideoPreview_Load(object sender, EventArgs e)
        {
            panel1.Width = (int)_window.GetInfo().WindowSize.Width;
            panel1.Height = (int)_window.GetInfo().WindowSize.Height;

            this.Width = panel1.Width + 43;
            this.Height = panel1.Height + 60;

            try
            {
                // Show the preview.
                _window.Show(true);
            }
            catch { }
        }

        /// <summary>
        /// Closing the form.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void VideoPreview_FormClosing(object sender, FormClosingEventArgs e)
        {
            try
            {
                _window.Show(false);
                _videoPreview.Stop();
            }
            catch { }
            DisposeCall();

            // Send the form closing event.
            OnVideoPreviewClosing?.Invoke(this, new EventArgs());
        }
    }
}
